crates/kubelet/src/plugin_watcher/mod.rs

```diff
@@ -0,0 +1,744 @@
+};
+
+use anyhow::Context;
+use log::{debug, error, trace, warn};
+use notify::Event;
+use tokio::fs::{create_dir_all, read_dir};
+use tokio::stream::StreamExt;
+use tokio::sync::RwLock;
+use tonic::Request;
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::path::{Path, PathBuf};
+
+#[cfg(target_family = "unix")]
+const DEFAULT_PLUGIN_PATH: &str = "/var/lib/kubelet/plugins_registry/";
+#[cfg(target_family = "windows")]
+const DEFAULT_PLUGIN_PATH: &str = "c:\\ProgramFiles\\kubelet\\plugins_registry";
+
+const SOCKET_EXTENSION: &str = "sock";
+
+/// An enum for capturing possible plugin types. This is purely for clarity and capturing this
+/// information is a compiled type as the information we get from gRPC is a string
+enum PluginType {
+    CSIPlugin,
+    DevicePlugin,
+}
+
+impl TryFrom<&str> for PluginType {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        match value {
+            "CSIPlugin" => Ok(PluginType::CSIPlugin),
+            "DevicePlugin" => Ok(PluginType::DevicePlugin),
+            _ => Err(anyhow::anyhow!(
+                "Unknown plugin type {}. Allowed types are 'CSIPlugin' and 'DevicePlugin'",
+                value
+            )),
+        }
+    }
+}
+
+/// Internal storage structure for a plugin
+#[derive(Debug)]
+struct PluginEntry {
+    plugin_path: PathBuf,
+    endpoint: Option<PathBuf>,
+}
+
+/// An internal storage plugin registry that implements most the same functionality as the [plugin
+/// manager](https://github.com/kubernetes/kubernetes/tree/fd74333a971e2048b5fb2b692a9e043483d63fba/pkg/kubelet/pluginmanager)
+/// in kubelet
+pub struct PluginRegistrar {
+    plugins: RwLock<HashMap<String, PluginEntry>>,
+    plugin_dir: PathBuf,
+}
+
+impl Default for PluginRegistrar {
+    fn default() -> Self {
+        PluginRegistrar {
+            plugin_dir: PathBuf::from(DEFAULT_PLUGIN_PATH),
+            plugins: RwLock::new(HashMap::new()),
+        }
+    }
+}
+
+impl PluginRegistrar {
+    /// Returns a new plugin registrar configured with the given plugin directory path
+    pub fn new<P: AsRef<Path>>(plugin_dir: P) -> Self {
+        PluginRegistrar {
+            plugin_dir: PathBuf::from(plugin_dir.as_ref()),
+            ..Default::default()
+        }
+    }
+
+    /// Gets the endpoint for the given plugin name, returning `None` if it doesn't exist
+    pub async fn get(&self, plugin_name: &str) -> Option<PathBuf> {
+        let lock = self.plugins.read().await;
+        lock.get(plugin_name)
+            .map(|v| v.endpoint.as_ref().unwrap_or(&v.plugin_path).to_owned())
+    }
+
+    /// Starts the plugin registrar and runs all automatic plugin discovery and registration loops
+    pub async fn run(&self) -> anyhow::Result<()> {
+        // Create plugin directory if it doesn't exist
+        create_dir_all(&self.plugin_dir).await?;
+
+        // Walk the plugin dir before hand and process any currently existing sockets
```

Comment:
Ok, so this is confusing. Perhaps I can add some documentation to clarify how Kubelet does this (because this is mostly from reverse engineering and gluing together disparate k8s docs). You basically have it right. The directory that kubelet is watching is where running plugins (which could be running in a container in normal k8s, but is more likely a separate service in Krustlet's case) place their sockets for Kubelet to communicate with. However, the actual specification of a `PluginInfo` supports two types of plugins: `CSIPlugin` and `DevicePlugin`. CSI plugins work by opening a `csi.sock` in another directory outside of the one watched by Kubelet. So there ends up being two endpoints: one for the plugin itself and another for the CSI stuff. DevicePlugins, which we don't yet support, only have the single path.

I'll add documentation to clarify this